New HSQL driver has dependency on Servlet

**Describe the bug**
The jdbc jar won't run without Servlet on the classpath.

**To Reproduce**
- Load the jdbc jar into DBeaver
- Run `select * from Patient`

See this stack
```
org.jkiss.dbeaver.model.sql.dbsqlexception: SQL Error: Internal jdbc driver errorat org.jkiss.dbeaver.model.impl.jdbc.exec.JDBCStatementImpl.executeStatement(JDBCStatementImpl.java:133)at org.jkiss.dbeaver.ui.editors.sql.execute.SQLQueryJob.executeStatement(SQLQueryJob.java:582)at org.jkiss.dbeaver.ui.editors.sql.execute.SQLQueryJob.lambda$1(SQLQueryJob.java:491)at org.jkiss.dbeaver.model.exec.DBExecUtils.tryExecuteRecover(DBExecUtils.java:189)at org.jkiss.dbeaver.ui.editors.sql.execute.SQLQueryJob.executeSingleQuery(SQLQueryJob.java:498)at org.jkiss.dbeaver.ui.editors.sql.execute.SQLQueryJob.extractData(SQLQueryJob.java:934)at org.jkiss.dbeaver.ui.editors.sql.SQLEditor$QueryResultsContainer.readData(SQLEditor.java:3928)at org.jkiss.dbeaver.ui.controls.resultset.ResultSetJobDataRead.lambda$0(ResultSetJobDataRead.java:123)at org.jkiss.dbeaver.model.exec.DBExecUtils.tryExecuteRecover(DBExecUtils.java:189)at org.jkiss.dbeaver.ui.controls.resultset.ResultSetJobDataRead.run(ResultSetJobDataRead.java:121)at org.jkiss.dbeaver.ui.controls.resultset.ResultSetViewer$ResultSetDataPumpJob.run(ResultSetViewer.java:5104)at org.jkiss.dbeaver.model.runtime.AbstractJob.run(AbstractJob.java:105)at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)Caused by: java.sql.SQLException: Internal jdbc driver errorat org.jkiss.dbeaver.model.impl.jdbc.exec.JDBCStatementImpl.handleExecuteError(JDBCStatementImpl.java:295)at org.jkiss.dbeaver.model.impl.jdbc.exec.JDBCStatementImpl.execute(JDBCStatementImpl.java:332)at org.jkiss.dbeaver.model.impl.jdbc.exec.JDBCStatementImpl.executeStatement(JDBCStatementImpl.java:131)... 12 moreCaused by: java.lang.NoClassDefFoundError: javax/servlet/ServletOutputStreamat ca.uhn.fhir.jpa.fql.jdbc.JdbcStatement.execute(JdbcStatement.java:125)at org.jkiss.dbeaver.model.impl.jdbc.exec.JDBCStatementImpl.execute(JDBCStatementImpl.java:330)... 13 moreCaused by: java.lang.ClassNotFoundException: javax.servlet.ServletOutputStreamat java.base/java.net.URLClassLoader.findClass(Unknown Source)at java.base/java.lang.ClassLoader.loadClass(Unknown Source)at java.base/java.lang.ClassLoader.loadClass(Unknown Source)... 15 more
```

**Expected behavior**
Get some results.

**Environment (please complete the following information):**
 - HAPI FHIR Version 6.8.0-snapshot

--- a/hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/6_8_0/5167-fix-jdbc-deps.yaml
+++ b/hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/6_8_0/5167-fix-jdbc-deps.yaml
@@ -1,0 +1,4 @@
+---
+type: fix
+issue: 5167
+title: "Fixed a depedency in the HSQL JDBC driver referencing a non-bundled class (javax.ServletOutputStream)"

--- a/hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/6_8_0/5167-fix-jdbc-deps.yaml
+++ b/hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/6_8_0/5167-fix-jdbc-deps.yaml
@@ -1,4 +1,4 @@
 ---
 type: fix
 issue: 5167
-title: "Fixed a depedency in the HSQL JDBC driver referencing a non-bundled class (javax.ServletOutputStream)"
+title: "Fixed a dependency in the HSQL JDBC driver referencing a non-bundled class (javax.ServletOutputStream)"

--- a/hapi-fhir-jpaserver-hfql/pom.xml
+++ b/hapi-fhir-jpaserver-hfql/pom.xml
@@ -32,6 +32,12 @@
 			<version>${project.version}</version>
 			<scope>test</scope>
 		</dependency>
+
+		<dependency>
+			<groupId>com.tngtech.archunit</groupId>
+			<artifactId>archunit-junit5</artifactId>
+			<scope>test</scope>
+		</dependency>
 	</dependencies>
 
 	<profiles>

--- a/hapi-fhir-jpaserver-hfql/src/main/java/ca/uhn/fhir/jpa/fql/jdbc/HfqlRestClient.java
+++ b/hapi-fhir-jpaserver-hfql/src/main/java/ca/uhn/fhir/jpa/fql/jdbc/HfqlRestClient.java
@@ -20,6 +20,7 @@
 package ca.uhn.fhir.jpa.fql.jdbc;
 
 import ca.uhn.fhir.jpa.fql.executor.IHfqlExecutionResult;
+import ca.uhn.fhir.jpa.fql.util.HfqlConstants;
 import ca.uhn.fhir.rest.client.impl.HttpBasicAuthInterceptor;
 import ca.uhn.fhir.util.IoUtil;
 import org.apache.commons.csv.CSVFormat;
@@ -27,8 +28,12 @@ import org.apache.commons.lang3.Validate;
 import org.apache.http.impl.client.CloseableHttpClient;
 import org.apache.http.impl.client.HttpClientBuilder;
 import org.apache.http.impl.conn.PoolingHttpClientConnectionManager;
+import org.hl7.fhir.r4.model.CodeType;
+import org.hl7.fhir.r4.model.IntegerType;
 import org.hl7.fhir.r4.model.Parameters;
+import org.hl7.fhir.r4.model.StringType;
 
+import javax.annotation.Nonnull;
 import java.sql.SQLException;
 import java.util.concurrent.TimeUnit;
 
@@ -65,6 +70,18 @@ public class HfqlRestClient {
 			httpClientBuilder.addInterceptorLast(new HttpBasicAuthInterceptor(theUsername, thePassword));
 		}
 		myClient = httpClientBuilder.build();
+	}
+
+	@Nonnull
+	public static Parameters newQueryRequestParameters(String sql, Integer limit, int fetchSize) {
+		Parameters input = new Parameters();
+		input.addParameter(HfqlConstants.PARAM_ACTION, new CodeType(HfqlConstants.PARAM_ACTION_SEARCH));
+		input.addParameter(HfqlConstants.PARAM_QUERY, new StringType(sql));
+		if (limit != null) {
+			input.addParameter(HfqlConstants.PARAM_LIMIT, new IntegerType(limit));
+		}
+		input.addParameter(HfqlConstants.PARAM_FETCH_SIZE, new IntegerType(fetchSize));
+		return input;
 	}
 
 	public IHfqlExecutionResult execute(

--- a/hapi-fhir-jpaserver-hfql/src/main/java/ca/uhn/fhir/jpa/fql/jdbc/HfqlRestClient.java
+++ b/hapi-fhir-jpaserver-hfql/src/main/java/ca/uhn/fhir/jpa/fql/jdbc/HfqlRestClient.java
@@ -33,9 +33,9 @@ import org.hl7.fhir.r4.model.IntegerType;
 import org.hl7.fhir.r4.model.Parameters;
 import org.hl7.fhir.r4.model.StringType;
 
-import javax.annotation.Nonnull;
 import java.sql.SQLException;
 import java.util.concurrent.TimeUnit;
+import javax.annotation.Nonnull;
 
 import static ca.uhn.fhir.jpa.fql.util.HfqlConstants.DEFAULT_FETCH_SIZE;
 import static org.apache.commons.lang3.ObjectUtils.defaultIfNull;

--- a/hapi-fhir-jpaserver-hfql/src/main/java/ca/uhn/fhir/jpa/fql/jdbc/JdbcStatement.java
+++ b/hapi-fhir-jpaserver-hfql/src/main/java/ca/uhn/fhir/jpa/fql/jdbc/JdbcStatement.java
@@ -20,7 +20,6 @@
 package ca.uhn.fhir.jpa.fql.jdbc;
 
 import ca.uhn.fhir.jpa.fql.executor.IHfqlExecutionResult;
-import ca.uhn.fhir.jpa.fql.provider.HfqlRestProvider;
 import ca.uhn.fhir.jpa.fql.util.HfqlConstants;
 import org.hl7.fhir.r4.model.Parameters;
 
@@ -43,8 +42,8 @@ class JdbcStatement implements Statement {
 	}
 
 	@Override
-	public ResultSet executeQuery(String sql) throws SQLException {
-		execute(sql);
+	public ResultSet executeQuery(String theSqlText) throws SQLException {
+		execute(theSqlText);
 		return getResultSet();
 	}
 
@@ -122,7 +121,7 @@ class JdbcStatement implements Statement {
 
 		int fetchSize = myFetchSize;
 
-		Parameters input = HfqlRestProvider.newQueryRequestParameters(sql, limit, fetchSize);
+		Parameters input = HfqlRestClient.newQueryRequestParameters(sql, limit, fetchSize);
 		IHfqlExecutionResult result = myConnection.getClient().execute(input, true, getFetchSize());
 
 		myResultSet = new JdbcResultSet(result, this);

--- a/hapi-fhir-jpaserver-hfql/src/main/java/ca/uhn/fhir/jpa/fql/provider/HfqlRestProvider.java
+++ b/hapi-fhir-jpaserver-hfql/src/main/java/ca/uhn/fhir/jpa/fql/provider/HfqlRestProvider.java
@@ -34,11 +34,11 @@ import org.apache.commons.csv.CSVPrinter;
 import org.hl7.fhir.instance.model.api.IPrimitiveType;
 import org.springframework.beans.factory.annotation.Autowired;
 
+import java.io.IOException;
+import java.io.OutputStreamWriter;
 import javax.annotation.Nullable;
 import javax.servlet.ServletOutputStream;
 import javax.servlet.http.HttpServletResponse;
-import java.io.IOException;
-import java.io.OutputStreamWriter;
 
 import static ca.uhn.fhir.jpa.fql.jdbc.HfqlRestClient.CSV_FORMAT;
 import static ca.uhn.fhir.rest.api.Constants.CHARSET_UTF8_CTSUFFIX;
@@ -210,5 +210,4 @@ public class HfqlRestProvider {
 			}
 		}
 	}
-
 }

--- a/hapi-fhir-jpaserver-hfql/src/main/java/ca/uhn/fhir/jpa/fql/provider/HfqlRestProvider.java
+++ b/hapi-fhir-jpaserver-hfql/src/main/java/ca/uhn/fhir/jpa/fql/provider/HfqlRestProvider.java
@@ -32,18 +32,13 @@ import ca.uhn.fhir.util.ValidateUtil;
 import ca.uhn.fhir.util.VersionUtil;
 import org.apache.commons.csv.CSVPrinter;
 import org.hl7.fhir.instance.model.api.IPrimitiveType;
-import org.hl7.fhir.r4.model.CodeType;
-import org.hl7.fhir.r4.model.IntegerType;
-import org.hl7.fhir.r4.model.Parameters;
-import org.hl7.fhir.r4.model.StringType;
 import org.springframework.beans.factory.annotation.Autowired;
 
-import java.io.IOException;
-import java.io.OutputStreamWriter;
-import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
 import javax.servlet.ServletOutputStream;
 import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.io.OutputStreamWriter;
 
 import static ca.uhn.fhir.jpa.fql.jdbc.HfqlRestClient.CSV_FORMAT;
 import static ca.uhn.fhir.rest.api.Constants.CHARSET_UTF8_CTSUFFIX;
@@ -124,6 +119,7 @@ public class HfqlRestProvider {
 				String continuation = toStringValue(theContinuation);
 				ValidateUtil.isTrueOrThrowInvalidRequest(
 						theOffset != null && theOffset.hasValue(), "No offset supplied");
+				@SuppressWarnings("java:S2259") // Sonar doesn't understand the above
 				int startingOffset = theOffset.getValue();
 
 				String statement = DatatypeUtil.toStringValue(theStatement);
@@ -147,6 +143,9 @@ public class HfqlRestProvider {
 				streamResponseCsv(theServletResponse, fetchSize, outcome, true, outcome.getStatement());
 				break;
 			}
+			default:
+				//noinspection DataFlowIssue
+				ValidateUtil.isTrueOrThrowInvalidRequest(false, "Unrecognized action: %s", action);
 		}
 	}
 
@@ -186,41 +185,30 @@ public class HfqlRestProvider {
 		theServletResponse.setContentType(CT_TEXT_CSV + CHARSET_UTF8_CTSUFFIX);
 		try (ServletOutputStream outputStream = theServletResponse.getOutputStream()) {
 			Appendable out = new OutputStreamWriter(outputStream);
-			CSVPrinter csvWriter = new CSVPrinter(out, CSV_FORMAT);
-			csvWriter.printRecords();
+			try (CSVPrinter csvWriter = new CSVPrinter(out, CSV_FORMAT)) {
+				csvWriter.printRecords();
 
-			// Protocol version
-			csvWriter.printRecord(HfqlConstants.PROTOCOL_VERSION, "HAPI FHIR " + VersionUtil.getVersion());
+				// Protocol version
+				csvWriter.printRecord(HfqlConstants.PROTOCOL_VERSION, "HAPI FHIR " + VersionUtil.getVersion());
 
-			// Search ID, Limit, Parsed FQL Statement
-			String searchId = theResult.getSearchId();
-			String parsedFqlStatement = "";
-			if (theInitialPage && theStatement != null) {
-				parsedFqlStatement = JsonUtil.serialize(theStatement, false);
+				// Search ID, Limit, Parsed FQL Statement
+				String searchId = theResult.getSearchId();
+				String parsedFqlStatement = "";
+				if (theInitialPage && theStatement != null) {
+					parsedFqlStatement = JsonUtil.serialize(theStatement, false);
+				}
+				csvWriter.printRecord(searchId, theResult.getLimit(), parsedFqlStatement);
+
+				// Print the rows
+				int recordCount = 0;
+				while (recordCount++ < theFetchSize && theResult.hasNext()) {
+					IHfqlExecutionResult.Row nextRow = theResult.getNextRow();
+					csvWriter.print(nextRow.getRowOffset());
+					csvWriter.printRecord(nextRow.getRowValues());
+				}
+				csvWriter.flush();
 			}
-			csvWriter.printRecord(searchId, theResult.getLimit(), parsedFqlStatement);
-
-			// Print the rows
-			int recordCount = 0;
-			while (recordCount++ < theFetchSize && theResult.hasNext()) {
-				IHfqlExecutionResult.Row nextRow = theResult.getNextRow();
-				csvWriter.print(nextRow.getRowOffset());
-				csvWriter.printRecord(nextRow.getRowValues());
-			}
-
-			csvWriter.close(true);
 		}
 	}
 
-	@Nonnull
-	public static Parameters newQueryRequestParameters(String sql, Integer limit, int fetchSize) {
-		Parameters input = new Parameters();
-		input.addParameter(HfqlConstants.PARAM_ACTION, new CodeType(HfqlConstants.PARAM_ACTION_SEARCH));
-		input.addParameter(HfqlConstants.PARAM_QUERY, new StringType(sql));
-		if (limit != null) {
-			input.addParameter(HfqlConstants.PARAM_LIMIT, new IntegerType(limit));
-		}
-		input.addParameter(HfqlConstants.PARAM_FETCH_SIZE, new IntegerType(fetchSize));
-		return input;
-	}
 }

--- a/hapi-fhir-jpaserver-hfql/src/test/java/ca/uhn/fhir/jpa/fql/jdbc/ArchitectureTest.java
+++ b/hapi-fhir-jpaserver-hfql/src/test/java/ca/uhn/fhir/jpa/fql/jdbc/ArchitectureTest.java
@@ -1,0 +1,27 @@
+package ca.uhn.fhir.jpa.fql.jdbc;
+
+import com.tngtech.archunit.core.domain.JavaClasses;
+import com.tngtech.archunit.core.importer.ImportOption;
+import com.tngtech.archunit.junit.AnalyzeClasses;
+import com.tngtech.archunit.junit.ArchTest;
+import com.tngtech.archunit.lang.syntax.ArchRuleDefinition;
+
+@AnalyzeClasses(
+	packages = "ca.uhn.fhir.jpa.fql..",
+	importOptions = {
+		ImportOption.DoNotIncludeTests.class
+	}
+)
+public class ArchitectureTest {
+	/**
+	 * This project has a "provided" dependency on javax.servlet, but the packaged jdbc driver doesn't bundle it.
+	 */
+	@ArchTest
+	void verifyNoDepsOnProvidedServlet(JavaClasses theJavaClasses) {
+
+		ArchRuleDefinition.noClasses().that().resideInAPackage("ca.uhn.fhir.jpa.fql.jdbc")
+			.should().transitivelyDependOnClassesThat().resideInAPackage("javax.servlet")
+			.check(theJavaClasses);
+	}
+
+}

--- a/hapi-fhir-testpage-overlay/src/main/java/ca/uhn/fhir/to/Controller.java
+++ b/hapi-fhir-testpage-overlay/src/main/java/ca/uhn/fhir/to/Controller.java
@@ -54,10 +54,6 @@ import org.springframework.web.bind.annotation.RequestMapping;
 import org.springframework.web.bind.annotation.RequestMethod;
 import org.springframework.web.bind.annotation.RequestParam;
 
-import javax.annotation.Nonnull;
-import javax.annotation.Nullable;
-import javax.servlet.ServletException;
-import javax.servlet.http.HttpServletRequest;
 import java.io.IOException;
 import java.io.StringWriter;
 import java.util.ArrayList;
@@ -65,6 +61,10 @@ import java.util.Collections;
 import java.util.List;
 import java.util.TreeSet;
 import java.util.stream.Collectors;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
 
 import static ca.uhn.fhir.rest.server.provider.ProviderConstants.DIFF_OPERATION_NAME;
 import static ca.uhn.fhir.util.UrlUtil.sanitizeUrlPart;
@@ -428,8 +428,7 @@ public class Controller extends BaseController {
 	protected IHfqlExecutionResult executeHfqlStatement(
 			HttpServletRequest theServletRequest, String theHfqlQuery, HomeRequest theRequest, int theRowLimit)
 			throws IOException {
-		Parameters requestParameters =
-				HfqlRestClient.newQueryRequestParameters(theHfqlQuery, theRowLimit, theRowLimit);
+		Parameters requestParameters = HfqlRestClient.newQueryRequestParameters(theHfqlQuery, theRowLimit, theRowLimit);
 		GenericClient client = theRequest.newClient(theServletRequest, getContext(theRequest), myConfig, null);
 		return new RemoteHfqlExecutionResult(requestParameters, client);
 	}

--- a/hapi-fhir-testpage-overlay/src/main/java/ca/uhn/fhir/to/Controller.java
+++ b/hapi-fhir-testpage-overlay/src/main/java/ca/uhn/fhir/to/Controller.java
@@ -6,9 +6,9 @@ import ca.uhn.fhir.context.RuntimeResourceDefinition;
 import ca.uhn.fhir.i18n.Msg;
 import ca.uhn.fhir.jpa.fql.executor.HfqlDataTypeEnum;
 import ca.uhn.fhir.jpa.fql.executor.IHfqlExecutionResult;
+import ca.uhn.fhir.jpa.fql.jdbc.HfqlRestClient;
 import ca.uhn.fhir.jpa.fql.jdbc.RemoteHfqlExecutionResult;
 import ca.uhn.fhir.jpa.fql.parser.HfqlStatement;
-import ca.uhn.fhir.jpa.fql.provider.HfqlRestProvider;
 import ca.uhn.fhir.model.api.Include;
 import ca.uhn.fhir.model.dstu2.valueset.ResourceTypeEnum;
 import ca.uhn.fhir.model.primitive.BoundCodeDt;
@@ -19,9 +19,17 @@ import ca.uhn.fhir.parser.DataFormatException;
 import ca.uhn.fhir.rest.api.Constants;
 import ca.uhn.fhir.rest.api.EncodingEnum;
 import ca.uhn.fhir.rest.client.impl.GenericClient;
+import ca.uhn.fhir.rest.gclient.ICreateTyped;
+import ca.uhn.fhir.rest.gclient.IHistory;
+import ca.uhn.fhir.rest.gclient.IHistoryTyped;
+import ca.uhn.fhir.rest.gclient.IHistoryUntyped;
+import ca.uhn.fhir.rest.gclient.IQuery;
+import ca.uhn.fhir.rest.gclient.IUntypedQuery;
 import ca.uhn.fhir.rest.gclient.NumberClientParam.IMatches;
+import ca.uhn.fhir.rest.gclient.QuantityClientParam;
 import ca.uhn.fhir.rest.gclient.QuantityClientParam.IAndUnits;
-import ca.uhn.fhir.rest.gclient.*;
+import ca.uhn.fhir.rest.gclient.StringClientParam;
+import ca.uhn.fhir.rest.gclient.TokenClientParam;
 import ca.uhn.fhir.rest.server.exceptions.BaseServerResponseException;
 import ca.uhn.fhir.to.model.HomeRequest;
 import ca.uhn.fhir.to.model.ResourceRequest;
@@ -46,6 +54,10 @@ import org.springframework.web.bind.annotation.RequestMapping;
 import org.springframework.web.bind.annotation.RequestMethod;
 import org.springframework.web.bind.annotation.RequestParam;
 
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
 import java.io.IOException;
 import java.io.StringWriter;
 import java.util.ArrayList;
@@ -53,14 +65,13 @@ import java.util.Collections;
 import java.util.List;
 import java.util.TreeSet;
 import java.util.stream.Collectors;
-import javax.annotation.Nonnull;
-import javax.annotation.Nullable;
-import javax.servlet.ServletException;
-import javax.servlet.http.HttpServletRequest;
 
 import static ca.uhn.fhir.rest.server.provider.ProviderConstants.DIFF_OPERATION_NAME;
 import static ca.uhn.fhir.util.UrlUtil.sanitizeUrlPart;
-import static org.apache.commons.lang3.StringUtils.*;
+import static org.apache.commons.lang3.StringUtils.defaultIfEmpty;
+import static org.apache.commons.lang3.StringUtils.defaultString;
+import static org.apache.commons.lang3.StringUtils.isBlank;
+import static org.apache.commons.lang3.StringUtils.isNotBlank;
 
 @org.springframework.stereotype.Controller()
 public class Controller extends BaseController {
@@ -418,7 +429,7 @@ public class Controller extends BaseController {
 			HttpServletRequest theServletRequest, String theHfqlQuery, HomeRequest theRequest, int theRowLimit)
 			throws IOException {
 		Parameters requestParameters =
-				HfqlRestProvider.newQueryRequestParameters(theHfqlQuery, theRowLimit, theRowLimit);
+				HfqlRestClient.newQueryRequestParameters(theHfqlQuery, theRowLimit, theRowLimit);
 		GenericClient client = theRequest.newClient(theServletRequest, getContext(theRequest), myConfig, null);
 		return new RemoteHfqlExecutionResult(requestParameters, client);
 	}

--- a/pom.xml
+++ b/pom.xml
@@ -1189,6 +1189,13 @@
 				<artifactId>jakarta.activation</artifactId>
 				<version>2.0.0</version>
 			</dependency>
+			<!-- Architecture Test -->
+			<dependency>
+				<groupId>com.tngtech.archunit</groupId>
+				<artifactId>archunit-junit5</artifactId>
+				<version>1.0.1</version>
+				<scope>test</scope>
+			</dependency>
 			<dependency>
 				<groupId>com.vladsch.flexmark</groupId>
 				<artifactId>flexmark</artifactId>

